rclcpp_action/src/server.cpp

```diff
@@ -12,30 +12,474 @@
+    publish_status();
+  }
+
+  // TODO(sloretz) make this fini happen in an exception safe way
+  ret = rcl_action_cancel_response_fini(&cancel_response);
+  if (RCL_RET_OK != ret) {
+    rclcpp::exceptions::throw_from_rcl_error(ret);
+  }
+
+  ret = rcl_action_send_cancel_response(
+    pimpl_->action_server_.get(), &request_header, response.get());
+  if (RCL_RET_OK != ret) {
+    rclcpp::exceptions::throw_from_rcl_error(ret);
+  }
+}
+
+void
+ServerBase::execute_result_request_received()
+{
+  rcl_ret_t ret;
+  // Get the result request message
+  rmw_request_id_t request_header;
+  std::shared_ptr<void> result_request = create_result_request();
+  ret = rcl_action_take_result_request(
+    pimpl_->action_server_.get(), &request_header, result_request.get());
+  if (RCL_RET_OK != ret) {
+    rclcpp::exceptions::throw_from_rcl_error(ret);
+  }
+
+  std::shared_ptr<void> result_response;
+
+  // check if the goal exists
+  std::array<uint8_t, 16> uuid = get_goal_id_from_result_request(result_request.get());
+  rcl_action_goal_info_t goal_info;
+  for (size_t i = 0; i < 16; ++i) {
+    goal_info.goal_id.uuid[i] = uuid[i];
+  }
+  if (!rcl_action_server_goal_exists(pimpl_->action_server_.get(), &goal_info)) {
+    // Goal does not exists
+    result_response = create_result_response(action_msgs::msg::GoalStatus::STATUS_UNKNOWN);
+  } else {
+    // Goal exists, check if a result is already available
+    auto iter = pimpl_->goal_results_.find(uuid);
+    if (iter != pimpl_->goal_results_.end()) {
+      result_response = iter->second;
+    }
+  }
+
+  if (result_response) {
+    // Send the result now
+    ret = rcl_action_send_result_response(
+      pimpl_->action_server_.get(), &request_header, result_response.get());
+    if (RCL_RET_OK != ret) {
+      rclcpp::exceptions::throw_from_rcl_error(ret);
+    }
+  } else {
+    // Store the request so it can be responded to later
+    auto iter = pimpl_->result_requests_.find(uuid);
+    if (iter == pimpl_->result_requests_.end()) {
+      pimpl_->result_requests_[uuid] = std::vector<rmw_request_id_t>{request_header};
+    } else {
+      iter->second.push_back(request_header);
+    }
+  }
+}
+
+void
+ServerBase::execute_check_expired_goals()
+{
+  // Allocate expecting only one goal to expire at a time
+  rcl_action_goal_info_t expired_goals[1];
+  size_t num_expired = 1;
+
+  // Loop in case more than 1 goal expired
+  while (num_expired > 0u) {
+    rcl_ret_t ret = rcl_action_expire_goals(
+      pimpl_->action_server_.get(), expired_goals, 1, &num_expired);
+    if (RCL_RET_OK != ret) {
+      rclcpp::exceptions::throw_from_rcl_error(ret);
+    } else if (num_expired) {
+      // A goal expired!
+      std::array<uint8_t, 16> uuid;
+      for (size_t i = 0; i < 16; ++i) {
+        uuid[i] = expired_goals[0].goal_id.uuid[i];
+      }
+      pimpl_->goal_results_.erase(uuid);
+      pimpl_->result_requests_.erase(uuid);
+    }
+  }
+}
+
+void
+ServerBase::publish_status()
```

Comment:
Just a thought:

IIRC, there was discussion in the design doc about making the status (and feedback) publisher optional. I'm not sure if we had any ideas about how the implementation would look, but given time constraints it's probably good to capture this feature in a ticket to tackle at a later time.
It might make sense to instead have the option to disable subscribers in the action client so then CLI tools can always interact with action servers.